### ISSUE TITLE
How to code SOS2 constraints in Julia with Gurobi solver 

### ISSUE THREAD

**Issue by mahsap58** [JuliaOpt/MathOptInterface.jl#978]
Hello, the following constraints are SOS2 constraints, please help me how to code these kine of constraints in Julia. My solver is gurobi. I tried to code but the error was: addSOS2 not defined.
![sos2](https://user-images.githubusercontent.com/57906598/71423179-cfe61d80-265d-11ea-8200-cabfdae622e6.png)

**Comment by odow** [comment]
In future, please post usage questions like this on our Discourse forum:
https://discourse.julialang.org/c/domain/opt/13

You're looking for
```Julia
using JuMP
model = Model()
@variable(model, x[1:4] >= 0)
@constraint(model, sum(x) == 1)
weights = [1.0, 2.0, 3.0, 4.0]
@constraint(model, x in MOI.SOS2(weights))
```
Note that `weights` defines the ordering of your variables in the SOS2 set.

**Comment by mahsap58** [comment]
I wrote the code like:
model = Model()
@variable(model, lambda[1:J,1:K,1:S] >= 0)
@constraint(model, con1[j = 1:J, s = 1:S], 1 == sum(lambda[j,k,s] for k=1:K))
weights = [1:J,1:K,1:S]
@constraint(model, x in MOI.SOS2(weights))

The error is:
LoadError: MethodError: no method matching MathOptInterface.SOS2(::Array{UnitRange{Int64},1})
Closest candidates are:
  MathOptInterface.SOS2(!Matched::Array{T<:Real,1}) where T<:Real

**Comment by odow** [comment]
You have multiple issues:
1) `x` was an example in my code. You need to use `lambda` instead.
2) I think you need an SOS constraint for each j and s
3) SOS constraints only apply to a vector of variables, not a matrix.

You need something like:
```Julia
model = Model()
J = K = S = 2
@variable(model, lambda[1:J, 1:K, 1:S] >= 0)
@constraint(model, [j = 1:J, s = 1:S], sum(lambda[j, k, s] for k = 1:K) == 1)
weights = collect(1:K)
@constraint(model, [j = 1:J, s = 1:S], lambda[j, :, s] in MOI.SOS2(weights))
```

**Comment by mahsap58** [comment]
I added these constraints to the code:
weight1 = collect(1:K)
@constraint(model, [j = 1:J, s = 1:S], phi[j, s, :] in MOI.SOS2(weight1))
weight2 = collect(1:S)
@constraint(model, [j = 1:J, k= 1:K], phi[j, :, k] in MOI.SOS2(weight2))
weight3 = collect(1:J)
@constraint(model, [s = 1:S, k= 1:K], phi[:, s, k] in MOI.SOS2(weight3))
the error is: 
LoadError: Gurobi.GurobiError(10005, "Unable to retrieve attribute 'ObjVal'")

**Comment by odow** [comment]
The error isn't related to the SOS constraints. It's likely that you have a bug in other aspects of your code (e.g., you're trying to access `objective_value(model)` when the model failed to solve correctly).

If you can't find the mistake, you should create a minimum working example (see https://discourse.julialang.org/t/psa-make-it-easier-to-help-you/14757), and then post on the Discourse forum: https://discourse.julialang.org/c/domain/opt/13.

Since this isn't a bug with MathOptInterface, I'm going to close this issue. Once you have an example, we can continue the discussion on Discourse.

**Comment by mahsap58** [comment]
Hello, 
Can you please help me. I am using the GLPKMathProgInterface pkg and face to Failed to precompile GLPKMathProgInterface error.please guide me what should I do about this issue.

Mahsa 

    On Thursday, December 26, 2019, 03:28:37 PM AST, Oscar Dowson <notifications@github.com> wrote:  
 
 
Closed #978.

**Comment by mlubin** [comment]
@mahsap58 Please avoid double posting. The right place for this discussion is https://discourse.julialang.org/t/the-error-of-using-glpkmathproginterface-pkg/34142.
